src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => delete stateCopy[key]);
+        break;
+
+      case 'clear':
+        stateCopy = {};
+        break;
+    }
+
+    stateHistory.push({ ...stateCopy });
```

Comment:
You are performing the action of copying the state object and pushing it into the stateHistory array in every case of the switch statement. To adhere to the DRY (Don't Repeat Yourself) principle, you should move this operation outside of the switch statement, after all cases, so it is executed just once after the switch.